CL2-6837-i2.1 BE Filter out folder admin_publications with no visible children

jinjagit opened the pull request:
## Checklist

- [ ] Added entry to changelog
  
NO CHANGELOG ENTRY - this PR does not fully implement any new feature.

- [x] Tests
<details>
<summary>More info</summary>

### Unit tests

Did you add relevant unit tests?

### E2E tests

Sometimes it can be more efficient to update E2E tests after CI has run them. If you know which ones to update, go ahead! E2E template cl2-back:

```bash
docker compose run --rm web bin/rails cl2_back:create_tenant[localhost,e2etests_template]
```

</details>

- [x] Prepared branch for code review
<details>
<summary>More info</summary>
Reviewed code to reduce unnecessary back and forth (removal of console.log, comments, ...)? Added comments to clarify code, emphasize what to pay attention to, etc.?
</details>

## Links

- [Jira ticket](https://citizenlab.atlassian.net/browse/CL2-6837)
- [Jira Epic](https://citizenlab.atlassian.net/browse/CL2-6832)

## What changes are in this PR?

Adds a filter, `'remove_childless_parents'` to `AdminPublicationsFilteringService`, that removes `admin_publications` that represent folders containing no visible projects.  

This is needed because the Epic adds positive matching of `admin_publications` with selected areas, and thus `admin_publications` representing empty folders, or folders that _appear_ empty due to the areas or topics filtering, should not be returned (since area matching is performed on projects contained in folders). It can also be used when filtering by topics (a feature to be added in this Epic)

### Example:
This request, _without_ the new query parameter, gets 3 `admin_publications` in the response;
```
http://localhost:3000/web_api/v1/admin_publications?areas%5B%5D=071a0213-d4ca-427c-ae4f-704274847938&publication_statuses%5B%5D=published&publication_statuses%5B%5D=archived&page%5Bnumber%5D=1&page%5Bsize%5D=6&remove_not_allowed_parents=true&depth=0

Response =>
{
    "data": [
        {
            "id": "5b323589-b42c-4557-92d8-77482253ce8d",
            "type": "admin_publication",
            "attributes": {
                "ordering": 1,
                "publication_status": "published",
                "depth": 0,
                "publication_title_multiloc": {
                    "en": "folder Champlinmouth project only",
                    "fr-BE": "folder Champlinmouth project only",
                    "nl-BE": "folder Champlinmouth project only"
                },
            ...
        },
        {
            "id": "5258304b-8aa7-40d5-97dc-29d2f78f7258",
            "type": "admin_publication",
            "attributes": {
                "ordering": 3,
                "publication_status": "published",
                "depth": 0,
                "publication_title_multiloc": {
                    "en": "folder Bechtelarshire project only",
                    "fr-BE": "folder Bechtalershire project only",
                    "nl-BE": "folder Bechtalershire project only"
                },
             ...
        },
        {
            "id": "7e5beaed-321e-452b-82bb-046d44604b54",
            "type": "admin_publication",
            "attributes": {
                "ordering": 4,
                "publication_status": "published",
                "depth": 0,
                "publication_title_multiloc": {
                    "en": "empty folder",
                    "fr-BE": "empty folder",
                    "nl-BE": "empty folder"
                },
            ...
        }
    ],
   ...
}
```
This request, which adds the new query parameter, `remove_childess_parents=true`, gets only the 1 `admin_publication` (a folder containing a project for the specified area) in the response;
```
http://localhost:3000/web_api/v1/admin_publications?areas%5B%5D=071a0213-d4ca-427c-ae4f-704274847938&publication_statuses%5B%5D=published&publication_statuses%5B%5D=archived&page%5Bnumber%5D=1&page%5Bsize%5D=6&remove_not_allowed_parents=true&remove_childless_parents=true&depth=0

Response =>
{
    "data": [
        {
            "id": "5b323589-b42c-4557-92d8-77482253ce8d",
            "type": "admin_publication",
            "attributes": {
                "ordering": 1,
                "publication_status": "published",
                "depth": 0,
                "publication_title_multiloc": {
                    "en": "folder Champlinmouth project only",
                    "fr-BE": "folder Champlinmouth project only",
                    "nl-BE": "folder Champlinmouth project only"
                },
            ...
        }
    ],
   ...
}
```

## How urgent is a code review?

By Wednesday, 10th November, so that FE work that will use this filtering can make use of this new functionality.

jinjagit committed:
Add filter and adjust related spec

jinjagit committed:
Merge branch 'master' of github.com:CitizenLabDotCo/citizenlab into CL2-6837-i2-1-remove-empty-folders-filter

jinjagit committed:
Test folder publication removed by filter - failing as expected

Currently, we are simply removing truly empty parent publications,
which equates to folders with no contents.
Now, we want to remove parent publications that contain only
children that should not be visible (e.g. filtered out by area).
This test tests if such a parent publication is removed, and fails
as we have not yet implemented this functionality.

jinjagit committed:
Don't list folders containing only projects for areas not selected

+ Improve tests (currently only acceptance tests for admin user)

jinjagit committed:
Remove redundant code

Now we are removing any folder publication that does not contain
at least one project for a selected area, there is no need to
remove empty folders in a separate filter.

jinjagit commented on back/app/services/admin_publications_filtering_service.rb:
```diff
@@ -5,6 +5,8 @@ class AdminPublicationsFilteringService
 
   # NOTE: This service is very fragile and the ORDER of filters matters for the Front-End, do not change it.
 
+  project_publications = nil
```
I'm not sure this is ideal. Suggestions to improve appreciated!

Because project_publications are found in the `'filter_projects'` filter, and I want to make use of this collection again in the `'remove_parents_with_only_not_allowed_children'` filter I add in this PR, I define this var up front to keep it available in all filter scopes ('scope' as in block-level/namespace scope).

adessy replied:
Sorry, but for me, this is a show-stopper because this will be a bug in a few months. The easiest would be to recompute the `project_publications` scope in the new filter. This should not change anything performance wise if the filtering is run as a single DB query.

jinjagit committed:
Re-order queries

adessy commented:
Hey Simon 👋 I had a quick look and before digging any further into the review, I'd like to challenge something: why not a "remove_empty_parents" filter? Do we really care that projects were removed by the area filtering?

jinjagit commented:
> why not a "remove_empty_parents" filter? Do we really care that projects were removed by the area filtering?

I kind of agree. I personally think it's probably better to never show empty folders (or folders that _appear to be_ empty), although I can also think of cases where customer might want to show such folders. But there are 2 issues here:
1. Product wants to show (and count) empty folders (when area or topic filtering _not_ used).
2. We need to filter out folders (when area filtering used) that do not contain any projects for the selected area(s), as in the [specs](https://citizenlab.atlassian.net/wiki/spaces/CL2/pages/167281035/Final+scope+specs). We can easily extend this approach to cover topics filtering too, also in the specs.
  
@juliennechen Correct? Or am I misrepresenting / misunderstanding Product view.

sebastienhoorens commented:
I agree that it would be better to try to keep the behaviour of the different filters as simple and as readable as possible. I feel like Adrien's solution could still work here (and I thought such a filter already existed, but doesn't seem to be the case).

1. Wouldn't this be solved by setting the filter to false? Or by leaving it out in the request parameters?
2. I would apply this filter after the other ones (this is one reason why I'm not so fond of the filterers). In this case, if the filter only applies after filtering out the projects that don't have the requested areas, I think you could still keep this parameter generic and leave out folders with no child projects left after the other filters were applied on those child projects. That way you can also get rid on the conditionality on the areas filter.

adessy commented:
Not entirely sure to follow, especially pt 2. But I'll try to clarify what I meant and see if we can get something out of it 🙂 

- What I had in mind is a filter that could be toggled using a new query parameter specific to that filter. It would make the application of the filter explicit and also more flexible. I think this addresses pt 1.
- By empty folders, I mean empty folders in the context of the filtered scope. They might contain projects, but those projects are not included in the scope at hand. If I understand pt 2 correctly, I think that's what you want, isn't it?

jinjagit commented:
> 1. Wouldn't this be solved by setting the filter to false? Or by leaving it out in the request parameters?

Not sure I understand. The filter has no specific related parameter to be left out, or to be set to `false`, (though I could rewrite it to use one - see response to 2). The filters are just applied in order, and if contain conditionals related to parameters are included in the filter method(s), then those are applied.

> 2. I would apply this filter after the other ones (this is one reason why I'm not so fond of the filterers). In this case, if the filter only applies after filtering out the projects that don't have the requested areas, I think you could still keep this parameter generic and leave out folders with no child projects left after the other filters were applied on those child projects. That way you can also get rid on the conditionality on the areas filter.

This is logical. I could add a parameter such as `remove_childless_parents`, and wrap the filter code in `if remove_childless_parents == true`, for example. The FE would need to use `remove_childless_parents=true` in the FE requests made when filtering `admin_publications` by area or topic filters.

In effect, we would remove conditionality on the `areas` param, and replace it with conditionality on another param. The FE will have one more param to use in some requests, but this can make use of the filter more explicit.

@adessy & @sebastienhoorens I think we have basically converged on an understanding here. I'll rewrite accordingly. Thanks! 🙏

jinjagit committed:
Make filter use explicit with use of specific query parameter

jinjagit commented:
@adessy @sebastienhoorens I've incorporated your suggestions 👍

adessy commented on back/app/services/admin_publications_filtering_service.rb:
```diff
@@ -51,4 +53,14 @@ class AdminPublicationsFilteringService
+  add_filter('remove_childless_parents') do |scope, options|
+    next scope unless ['true', true, '1'].include? options[:remove_childless_parents]
+
+    parents_of_project_publications_ids = project_publications.where.not(parent_id: nil).map(&:parent_id)
```
The `map` will trigger a DB query here.

jinjagit replied:
I've replaced it, using your suggestion with `select`. Does this mean we should also edit out the existing `map` statements in the first `'remove_not_allowed_parents'` filter?

adessy replied:
That's up to you 😉

jinjagit replied:
I tried replacing the other uses of `map`, but everything in `admin_publications_filtering_spec` failed, and I think I would need to understand `class MockAdminPublicationsTree` before I could hope to understand why. So I'll leave it for another day / ticket.

adessy commented on back/app/services/admin_publications_filtering_service.rb:
```diff
@@ -51,4 +53,14 @@ class AdminPublicationsFilteringService
+    parents_of_project_publications     = scope.where(id: parents_of_project_publications_ids)
+    non_parents                         = scope.where(children_allowed: false)
+    
+    parents_of_project_publications.or(non_parents)
```
What do you think about?
```suggestion
    parents_with_children = scope.where(id: scope.select(:parent_id).where.not(parent_id: nil).distinct)
    non_parents = scope.where(children_allowed: false)
    parents_with_children.or(non_parents)
```

adessy commented on back/spec/acceptance/admin_publications_spec.rb:
```diff
@@ -108,6 +110,44 @@
+          p1.save!
+
+          do_request(areas: [a2.id], remove_childless_parents: true)
+          json_response = json_parse(response_body)
```
(there exists `json_response_body` helper)

adessy commented on back/spec/acceptance/admin_publications_spec.rb:
```diff
@@ -108,6 +110,44 @@
+          do_request(areas: [a2.id], remove_childless_parents: true)
+          json_response = json_parse(response_body)
+
+          expect(json_response[:data].map { |d| d.dig(:relationships, :publication, :data, :id) }).to include @folder.id
```
(even better, `response_data` helper)

adessy commented on back/spec/acceptance/admin_publications_spec.rb:
```diff
@@ -108,6 +110,44 @@
+
+          p3 = @projects[2]
+          p3.areas << a1
+          p3.save!
```
Do you really think it's worth the trouble of adding an extra area to the projects? If so, I wouldn't do it for the 3 projects then, to spare a few db queries and make the test faster.

jinjagit replied:
The reason I do it for all 3 projects in the folder is because currently we treat a project with no area selected as having _all_ areas selected. My next PR in this Epic should remove that rule, and then I could take this test down to using only one of these projects with a specified area.

adessy replied:
Alright, makes sense. May I suggest iteration then 🙂

jinjagit replied:
You may :monocle_face:

adessy replied:
Something like `@folder.projects.each {|p| p.update(areas: [a1]) }`.

adessy reviewed (changes requested):
Hey Simon 👋 I don't approve this one just yet because of the "hidden" shared state introduced by `project_publications = nil`. I put a few suggestions for this and a few other minor things in comments. Tell me what you think!

jinjagit commented:
> Hey Simon 👋 I don't approve this one just yet because of the "hidden" shared state introduced by `project_publications = nil`. I put a few suggestions for this and a few other minor things in comments. Tell me what you think!

That's great! I had a bad feeling about the shared state, which is why I highlighted my concerns in the code comment. Glad to hear I can recompute the scope in the filter without performance hit :+1:

All your comments are helpful. I will work on incorporating your advice today. Thanks again!

jinjagit committed:
Remove use of hidden shared state

We can simply reuse the filtering to get project_publications.
Because this part of one query, it should not add any performace
hit.

jinjagit committed:
Bugfix: Don't exclude parents of visible projects

We need to ensure we get all admin_publications representing
projects that are visible to user, and to do this we need to
ensure we don't use the top_level_only (depth=0) filtered scope
as a starting point, if this parameter has been used. This means
project_publications will include projects in folders.

Need to create a test to catch this, as current specs were all
passing!

jinjagit committed:
Move filter to before 'top_level_only' filter

This removes need to add new queries to get project publications
of projects in folders, since depth: 0 could not be applied yet
and thus scope will still include such child project publications.

jinjagit committed:
Simplify query and use select not map

Now we have moved the filter to before the depth: 0 filter, we
can be sure that all filtered project publications are still
present, including the children that we use to filter out
folder publications with no visible children.

jinjagit commented:
@adessy  I _eventually_ realised that by moving the filter to _before_ the filter that uses `depth: 0`, I completely remove the reason for my hidden state variable, and also could then use your simplified query which uses `scope`. I no longer need to get the filtered projects again, because I no longer need to replace the project publications removed by the `depth: 0` filter! 🤦

jinjagit committed:
Use response_data spec helper

jinjagit committed:
Relace map with select in 'remove_not_allowed_parents' filter

Prevents breaking the current query into multiple queries.

jinjagit committed:
Use each block to reduce repetition in spec

jinjagit committed:
Revert changes to 'remove_not_allowed_parents' filter

adessy commented on back/spec/acceptance/admin_publications_spec.rb:
```diff
@@ -108,6 +110,32 @@
+          expect(response_data.map { |d| d.dig(:relationships, :publication, :data, :id) }).to include @folder.id
+        end
+
+        example "Don't list admin publications representing folders that don't contain any project(s) with the specified areas;" do
```
(pico: why ';' at the end? 🙂)

jinjagit replied:
No idea 😄 Will remove.

nTraum commented on back/spec/acceptance/admin_publications_spec.rb:
```diff
@@ -31,9 +31,11 @@
       parameter :areas, 'Filter by areas (AND)', required: false
-      parameter :folder, "Filter by folder (project folder id)", required: false
       parameter :publication_statuses, "Return only publications with the specified publication statuses (i.e. given an array of publication statuses); always includes folders; returns all publications by default", required: false
-      parameter :remove_childless_parents, 'Use the visibility rules of children on the parent and remove the empty ones', required: false
```
I'm confused why this parameter was already present before this change. To me the diff in `back/app/services/admin_publications_filtering_service.rb` looks like we just introduced it in this PR? :thinking:

jinjagit replied:
Yes. I believe a filter with the same name had existed at some time in the past, then was removed but the parameter declaration in the test was not (easy to miss as doesn't cause any issues). I then recreated a filter using the same name, and therefore left the parameter declaration in the test.

nTraum commented on back/app/services/admin_publications_filtering_service.rb:
```diff
@@ -48,6 +48,15 @@ class AdminPublicationsFilteringService
+    next scope unless ['true', true, '1'].include? options[:remove_childless_parents]
+
+    parents_with_children = scope.where(id: scope.select(:parent_id).where.not(parent_id: nil).distinct)
+    non_parents           = scope.where(children_allowed: false)
```
Are `non_parents` the same as `children` or are they something different semantically? If they are the same, I'd find `children` as a variable name more intention-revealing.

jinjagit replied:
`non_parents` and `children` are, indeed, different semantically: `children` are items with a parent (they are in a folder). `non_parents` includes items with no children allowed, and possibly no parent (e.g. top-level projects not in folders).

nTraum reviewed (approved):
Code looks great :rocket: . Left some questions that would help my understanding as a newbie a bit. :grimacing:

jinjagit committed:
Merge branch 'master' of github.com:CitizenLabDotCo/citizenlab into CL2-6837-i2-1-remove-empty-folders-filter

jinjagit committed:
Remove unnecessary semi-colons

jinjagit committed:
Merge branch 'master' into CL2-6837-i2-1-remove-empty-folders-filter
